components/proxy/src/main/java/com/hotels/styx/spi/ObjectFactories.java

```diff
@@ -53,60 +52,44 @@ private ObjectFactories() {
     }
 
     private static <T> Optional<T> newInstance(Class<T> type, String className, Optional<Path> pathOptional) {
-        Iterable<ClassLoader> classLoaders = pathOptional
-                .map(path -> makeClassLoadersFor(readResourcesFrom(path)))
-                .orElseGet(() -> singleton(ObjectFactories.class.getClassLoader()));
-
-        List<T> instances = newArrayList();
-        for (ClassLoader classLoader : classLoaders) {
-            try {
-                Class<?> aClass = Class.forName(className, false, classLoader);
-                T instance = type.cast(aClass.newInstance());
-                instances.add(instance);
-            } catch (ClassNotFoundException e) {
-                LOGGER.debug(e.getMessage(), e);
-            } catch (InstantiationException | IllegalAccessException e) {
-                LOGGER.error("error loading class={}", className);
-                return Optional.empty();
-            }
-        }
-
-        int size = instances.size();
-        if (size == 0) {
+        ClassLoader classLoader = pathOptional
+                .map(path -> classLoader(readResourcesFrom(path)))
+                .map(ClassLoader.class::cast)
+                .orElseGet(ObjectFactories.class::getClassLoader);
+
+        try {
+            Class<?> aClass = Class.forName(className, false, classLoader);
+            T instance = type.cast(aClass.newInstance());
+            return Optional.of(instance);
+        } catch (ClassNotFoundException e) {
             LOGGER.error("no class={} is found in the specified classpath={}", className, pathOptional.orElse(Paths.get("")));
             return Optional.empty();
+        } catch (InstantiationException | IllegalAccessException e) {
+            LOGGER.error("error loading class={}", className);
```

Comment:
I'm not too sure that using Optionals here is the best option. Right now, when we return empty we just throw an Exception, but that Exception doesn't include the real cause, which we are logging separatedly. Isn't it easier to just propagate the Exception and log the full error? 
Thoughts?